pallets/assets-registry/src/lib.rs

```diff
@@ -183,6 +183,94 @@ pub mod pallet {
 		}
 	}
 
+	// Should adapter the representation of asset location after reanchored.
+	// Because xcm would reanchore the location of the asset that reserved on our chain.
+	// https://github.com/paritytech/polkadot/pull/4470
+	pub trait ReserveAssetChecker {
+		// Return true if asset is reserved on local
+		fn is_reserve_asset(asset: &MultiAsset) -> bool;
+		// Return true if given location is reserved on local
+		fn is_reserve_asset_location(id: &MultiLocation) -> bool;
+		// Return asset origin location if asset is reserved on local, otherwise return None
+		fn to_absoluted_location(location: &MultiLocation) -> Option<MultiLocation>;
+		// Return a new asset with a absoluted location if asset is reserved on local.
+		fn to_noncanonical_asset(asset: &MultiAsset) -> MultiAsset;
+	}
+
+	pub struct ReserveAssetFilter<T, I>(PhantomData<(T, I)>);
+	impl<T: Get<ParaId>, I: NativeAssetChecker> ReserveAssetChecker for ReserveAssetFilter<T, I> {
+		fn is_reserve_asset(asset: &MultiAsset) -> bool {
+			if I::is_native_asset(asset) {
+				return true;
+			} else {
+				return match &asset.id {
+					Concrete(ref id) if Self::is_reserve_asset_location(id) => true,
+					_ => false,
+				};
+			}
+		}
+
+		fn is_reserve_asset_location(id: &MultiLocation) -> bool {
+			if I::is_native_asset_location(id) {
+				return true;
+			} else {
+				return match id.reserve_location() {
+					Some(reserve_location) => {
+						return match (reserve_location.parents, reserve_location.first_interior()) {
+							// Any assets match our parachain id
+							(1, Some(para)) => para == &Parachain(T::get().into()),
+							// ChainBridge assets with noncanonical location
+							(0, Some(GeneralKey(cb_key))) => cb_key == CB_ASSET_KEY,
+							// (0, Here) is expressed as PHA location
+							(0, None) => true,
+							_ => false,
+						};
+					}
+					_ => false,
+				};
+			}
```

Comment:
All the `return` can be omitted

Reply:
Btw, now we can normalize any location to the canonical location. Do you think we can just normalize it first, and then we don't need to care about `../2004` specially?

Reply:
Yes, since `is_native_asset_location()` matches the PHA location and `reserve_location()` always returns canonical location if asset is reserved on our side, we can freely remove `../2004` and `here`.